VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -343,75 +354,135 @@ void CookVitruvioActors(TArray<AActor*> Actors)
+							}
+						}
+						else
+						{
+							FName Name = MakeUniqueObjectName(CookedActor, UHierarchicalInstancedStaticMeshComponent::StaticClass(), *InstanceMesh->GetName());
+							auto* InstancedStaticMeshComponent = AttachMeshComponent<UHierarchicalInstancedStaticMeshComponent>(CookedActor, CookedMeshComponent, InstanceMesh, Name, GeneratedModelHismComponent->GetComponentTransform());
+
+							for (int32 InstanceIndex = 0; InstanceIndex < GeneratedModelHismComponent->GetInstanceCount(); ++InstanceIndex)
+							{
+								FTransform Transform;
+								GeneratedModelHismComponent->GetInstanceTransform(InstanceIndex, Transform);
+								InstancedStaticMeshComponent->AddInstance(Transform);
+							}
+						}
 					}
 				}
 			}
-
+			
 			FString OldActorLabel = Actor->GetActorLabel();
 
-			// Destroy the old procedural Vitruvio Actor
-			Actor->Destroy();
+			// Destroy the old procedural Vitruvio Actors
+			if (VitruvioBatchActor)
+			{
+				TSet<UVitruvioComponent*> BatchedVitruvioComponents = VitruvioBatchActor->GetVitruvioComponents();
+				VitruvioBatchActor->UnregisterAllVitruvioComponents();
+				for (UVitruvioComponent* BatchedVitruvioComponent : BatchedVitruvioComponents)
+				{
+					BatchedVitruvioComponent->GetOwner()->Destroy();
+				}
+			}
+			else
+			{
+				Actor->Destroy();
+			}
+			
 
 			CookedActor->SetActorLabel(OldActorLabel);
 
 			GEditor->SelectActor(CookedActor, true, false);
 		}
 
 		IsCooking = false;
-		VitruvioModule::Get().OnAllGenerateCompleted.Remove(ModelsGeneratedHandle);
-		ModelsGeneratedHandle.Reset();
-	});
+	};
 
 	// Regenerate the selected Actors to make sure we have a model to cook.
+	TArray<UVitruvioComponent*> VitruvioComponents;
 	for (AActor* Actor : Actors)
 	{
-		UVitruvioComponent* VitruvioComponent = Actor->FindComponentByClass<UVitruvioComponent>();
-		if (!VitruvioComponent)
+		if (UVitruvioComponent* VitruvioComponent = Actor->FindComponentByClass<UVitruvioComponent>())
+		{
+			VitruvioComponents.Add(VitruvioComponent);
+		}
+		else if (AVitruvioBatchActor* VitruvioBatchActor = Cast<AVitruvioBatchActor>(Actor))
 		{
-			continue;
+			UGenerateCompletedCallbackProxy* CallbackProxy = NewObject<UGenerateCompletedCallbackProxy>();
+			CallbackProxy->RegisterWithGameInstance(VitruvioBatchActor);
+			CallbackProxy->OnGenerateCompleted.AddLambda([DoCook] {
+				DoCook();
+			});
+			
+			VitruvioBatchActor->GenerateAll(CallbackProxy);
 		}
+	}
 
-		VitruvioComponent->Generate();
+	if (VitruvioComponents.Num() > 0)
+	{
+		UGenerateCompletedCallbackProxy* CallbackProxy = NewObject<UGenerateCompletedCallbackProxy>();
+		CallbackProxy->RegisterWithGameInstance(VitruvioComponents[0]);
+		CallbackProxy->OnGenerateCompleted.AddLambda(FExecuteAfterCountdown(VitruvioComponents.Num(), [DoCook]() {
+			DoCook();
+		}));
+	
+		for (UVitruvioComponent* VitruvioComponent : VitruvioComponents)
+		{
+			VitruvioComponent->Generate(CallbackProxy);
+		}

```

Comment:
I think in some cases (e.g. if there is a UVitruvioComponent & AVitruvioBatchActor) the DoCook lambda is executed twice instead of just once.

Reply:
You are right. Refactored the code and made sure that it's only called once (and is also more readable).